README
Finance-QuoteHist 
-----------------

The Finance-QuoteHist bundle is several modules designed to fetch
historical stock quotes from the web: 
  
Finance::QuoteHist
------------------
Top level aggregator that will select a default lineup of
site instances from which to retrieve quotes. Other
than the default lineup, this module behaves identically
to whichever site-specific module is first in the lineup.
See below for site-specific modules. 

Finance::QuoteHist::Generic
---------------------------
The real workhorse of the bundle, a sub class of
LWP::UserAgent. Site-specific modules are sub classes
of this module.

Current site-specific modules include:

Finance::QuoteHist::Yahoo
Finance::QuoteHist::Yahoo::Australia
Finance::QuoteHist::QuoteMedia
Finance::QuoteHist::Google
Finance::QuoteHist::BusinessWeek
Finance::QuoteHist::MSN


INSTALLATION

You install HTML-Element-Extended, as you would install any
perl library, by running these commands:

   perl Makefile.PL
   make
   make test
   make install


DOCUMENTATION

POD style documentation is included with each module. This is normally
converted to a manual page and installed as part of the "make install"
process. You should also be able to use the 'perldoc' utility to
extract and read documentation from the module file directly.  See
Changes for recent changes.


SUPPORT

There is a mailing list available. To view past messages or subscribe,
go to:

   http://lists.sourceforge.net/mailman/listinfo/quotehist-general

Questions and comments can also be directed to Matt Sisk <sisk@mojotoad.com>

AVAILABILITY

The library is available from CPAN:

   http://www.cpan.org/authors/id/M/MS/MSISK/

The latest version is also available at:

   http://www.mojotoad.com/sisk/projects/Finance-QuoteHist/


DISCLAIMER 

The data returned from these modules is in no way guaranteed, nor
are the developers responsible in any way for how this data (or lack
thereof) is used. The interface is based on URLs and page layouts
that might change at any time. Even though these modules are designed
to be adaptive under these circumstances, they will at some point
probably be unable to retrieve data unless fixed or provided with new
parameters. Furthermore, the data from these web sites is usually not
even guaranteed by the web sites themselves, and oftentimes is
acquired elsewhere. See the documentation for each site-specific
module for more information regarding the disclaimer for that site. 


ACKNOWLEDGEMENTS

Thanks to Jim Miner for his generous efforts testing this module. Thanks
to Josh Woodward for some pointers on the Yahoo! interface. Thanks to
Jacob Anawalt for pointers on the Yahoo! interface and suggestions on
open-ended date queries. Thanks to Paul McDermott and Jay Strauss for
patches. Thanks to Manoj Bhatti, Bernd Hacker, Christian von
Engelbrechten, Gary Nielson, Mike Brown, Bill Stephenson, and Robert S.
Weigel for their bug-spotting.


COPYRIGHT

Copyright (c) 1999-2009 Matthew P. Sisk.  All rights reserved.
All wrongs revenged. This program is free software; you can
redistribute it and/or modify it under the same terms as
Perl itself.


Changes
Revision history for Perl extension Finance::QuoteHist.

1.18  Mon Aug 29 10:44:57 EDT 2011
    - fixed test data window (1.17 was a little *too*
      recent for reliable monthly tests)

1.17  Fri Aug 26 19:31:05 EDT 2011
    - Updated tests with more recent data

1.16  Thu Sep 30 01:00:27 EDT 2010
    - Fixed tests so that they no longer rely upon historical adjusted
      quotes since these appear to change every now and then whenever
      Yahoo decides to tweak their algorithm.
    - typo fix

1.14  Mon Jun  7 18:36:30 EDT 2010
    - Fixed several syntax issues that were tripping up -w
    - Fixed the utf8 Byte Order Mark issues with google csv
    - Added modules for Investopedia.com and StockNod.com
    - Expanded site-specific tests (for development mostly)
    - Test data updated, docs refreshed and tweaked

1.12  Mon Aug  3 18:40:15 EDT 2009
    - Fixed various site-specific modules
    - F::Q::Yahoo_AU is deprecated in favor of
      F::Q::Yahoo::Australia (though the old package
      is still around for backwards compat)
    - moved CSV class discovery to package level (thanks
      to Edmond Shum)
    - fixed the 'unsigned integer' problem with volumes
      (thanks to 'o kite')
    - fixed some undef warnings and typos (thanks to
      Jim Miner)

1.11  Mon Jun 25 19:21:21 EDT 2007
    - Added Google and Yahoo_AU
    - Rewrote BusinessWeek, decodes from javascript channels now,
      since they changed how their site operates. No more intraday
      quotes though.

1.10  Tue Mar 13 13:57:07 EDT 2007

    - Tweaked Yahoo test data yet again (they seem to be rounding off
      their volumes now)
    - Fixed undefined error for symbols with no closing (thanks to
      Terry Gliedt)
    - Added BusinessWeek to lineup, including intraday functionality.
    - Flattened champion delegation.
    - Added labels() documentation
    - Implemented user-defined row filtering callback (thanks to Manoj
      Bhatti for the suggestion)
    - Added 'shares' as possible label for 'vol' (thanks again to Manoj)

1.09  Thu Jun  8 16:19:57 EDT 2006
    - Fixed ^DJI test data (again) since yahoo keeps changing how they
      represent volume (thanks to Ivo Welch)

1.08  Sat May  6 15:16:11 EDT 2006
    - Fixed ^DJI test data for quote tests (Yahoo switched to the volume
      for each anchor date rather than aggregate volumes)
    - splits() now returns ref properly in scalar context for
      F::QH::Yahoo. Thanks to Raymond de Leon for spotting.

1.07  Thu Mar 16 00:44:24 EST 2006
    - Internal restructure to date iterators where appropriate.
    - Doc fix for MSN.
    - Various minor edge cases fixed.

1.06  Fri Mar 10 15:59:05 EST 2006
    - Added MSN quote source.
    - Fixed some edge cases involving null-queries on splits.
    - Various minor fixes, including proper business-day calculations
      for end of month dates on montly resolutions, plus proper newline
      stripping (chomp didn't work on MSN csv).

1.05  Tue Feb 28 16:47:57 EST 2006
    - Splits now have no dependencies on dividend queries
    - Daily, Weekly, Monthly granularities available
    - Restored quote_source(), dividend_source(), and
      split_source() methods
    - More tests added

1.04  Fri Jan  6 12:18:59 EST 2006
    - Compensated for yahoo symbol switch from ^DJA to ^DJI, which was
      causing some tests to fail.
    - Updated quotemedia url.

1.03  Wed Nov  2 16:52:30 EST 2005
    - Minor bug fixes
    - Forced date calcs to GMT to avoid some cygwin/win errors
    - Added auto_proxy, enabled during tests so that if $ENV{HTTP_PROXY}
      is present $ua->env_proxy will be automatically invoked
    - Tests will skip if network connectivity cannot be established

1.02  Wed Aug  3 14:46:43 EDT 2005
    - Fixed Date::Manip requirement that was inadvertently dropped.

1.01  Thu Jul 21 13:29:54 EDT 2005
    - Added conditional support for Text::CSV_PP in cases where
      Text::CSV_XS is not installed
    - Split out tests

1.00  Thu Jun 23 16:25:53 EDT 2005
    - Reworked Yahoo! queries for new interface
    - Added quotes from QuoteMedia (they supply SiliconInvestor, so that
      was dropped)
    - Date constraints relaxed. No end date defaults to 'today', no
      start date defaults to inception. Having neither date grabs
      entire history.
    ! Internal API changed from symbolic dereferencing to supplying
      closures and iterators for particular targets and parse modes.

0.33
    - Broke up Yahoo! csv queries into blocks of 1000, as apparently
      Yahoo! sometimes puts a cap on these sort of queries (thanks to
      Jay Strauss)

0.32  Thu Aug 14 13:37:36 EDT 2003
    - Fixed csv parsing for Yahoo dividend data after they fixed
      their format
    - Updated test data

0.31  Thu Jan 16 19:20:31 GMT 2003
    - Rolled back some 5.6 specific syntax

0.30  Fri Nov  8 16:47:07 EST 2002
    - adjusted volume fix (column labeling snafu)
    - extraneous status messages fixed for split extrations
    - added filtered split and filtered dividend tests

0.29  Thu Oct 17 04:36:05 EDT 2002
    - Patched Yahoo URL interface since they switched to January = 00
      rather than 01.

0.28  Thu Feb 28 08:23:08 CST 2002
    - Updated Yahoo interface, incorporating new changes.
    - added proxy_env parameter and method to ease proxy handling via
      LWP::UserAgent
    - Added WallStreetCity.pm (basic source)
    - Removed FinancialWeb.pm (apparently defunct)
    - Removed MotleyFool.pm (no longer offers tabular data, plus login
      now required for historical info)

0.25  Mon Jan  8 19:35:00 CST 2001
    - Doc cleanup

0.24  Wed Nov 29 02:28:59 CST 2000
    - Replaced FinancialWeb with SiliconInvestor in order to provide
      information on defunct ticker symbols -- FinancialWeb ceased
      coverage of these symbols.

0.23  Thu Nov  2 15:19:00 CST 2000
    - Fixed cache bug with Yahoo::source_type()
    - Documentation revisions and fixes

0.21  Wed Sep 13 13:25:35 CDT 2000

lib/Finance/QuoteHist/Yahoo.pm
package Finance::QuoteHist::Yahoo;

use strict;
use vars qw(@ISA $VERSION);
use Carp;

$VERSION = '1.05';

use Finance::QuoteHist::Generic;
@ISA = qw(Finance::QuoteHist::Generic);

use HTML::TableExtract 2.07;
use Date::Manip;
Date::Manip::Date_Init("TZ=GMT");

# Example for HTML output:
#
# http://finance.yahoo.com/q/hp?s=IBM&a=00&b=2&c=1800&d=04&e=8&f=2005&g=d&z=66&y=66
#
#   * s - ticker symbol
#   * a - start month
#   * b - start day
#   * c - start year
#   * d - end month
#   * e - end day
#   * f - end year
#   * g - resolution (e.g. 'd' is daily, 'w' is weekly, 'm' is monthly)
#   * y is the offset (cursor) from the start date
#   * z is the number of results to return starting at the cursor (66
#     maximum, apparently)
#
# Note alternate url:
# http://table.finance.yahoo.com/d?a=1&b=1&c=1800&d=3&e=1&f=2006&s=yhoo&y=200&g=d
#
# Example for CSV output:
#
# http://ichart.finance.yahoo.com/table.csv?s=IBM&a=00&b=2&c=1800&d=04&e=8&f=2005&g=d&ignore=.csv
#
# (historically could use table.finance.yahoo.com as well)
#
# Note that Yahoo implements month numbering with Jan=0 and Dec=11.
#
# For CSV output, date ranges are unlimited; the output is adjusted and
# does not include any split or dividend notices.
#
# URL for splits and dividends:
#
# These are either extracted from within the historical quote results
# (non_quote_row()) or found directly:
#
# http://finance.yahoo.com/q/hp?s=IBM&a=00&b=2&c=1800&d=04&e=8&f=2005&g=v
# http://table.finance.yahoo.com/table.csv?s=IBM&a=00&b=2&c=1800&d=04&e=8&f=2005&g=v&ignore=.csv
#
# Example URL for weekly:
#
# http://finance.yahoo.com/q/hp?s=IBM&a=00&b=2&c=1962&d=01&e=27&f=2006&g=w
#
# Example URL for monthly:
#
# http://finance.yahoo.com/q/hp?s=IBM&a=00&b=2&c=1962&d=01&e=27&f=2006&g=m

sub new {
  my $that = shift;
  my $class = ref($that) || $that;
  my %parms = @_;

  $parms{parse_mode} ||= 'csv';

  my $self = __PACKAGE__->SUPER::new(%parms);
  bless $self, $class;

  $self->set_label_pattern(
    target_mode => 'dividend',
    parse_mode  => 'html',
    label       => 'div',
    pattern     => qr/(Open|Div)/
  );

  $self;
}

# override these for other Yahoo sites
sub url_base_csv    { 'http://ichart.finance.yahoo.com/table.csv' }
sub url_base_html   { 'http://finance.yahoo.com/q/hp' }
sub url_base_splits { 'http://finance.yahoo.com/q/bc' }

sub granularities { qw( daily weekly monthly ) }

# Yahoo can fetch dividends and splits. They can be extracted from
# regular quote results or queried directly.

sub html_parser {
  my($self, %parms) = @_;
  my $target_mode = $parms{target_mode} || $self->target_mode;
  return $self->SUPER::html_parser(%parms) unless $target_mode eq 'split';
  sub {
    my $data = shift;
    my $html_string;
    if (ref $data) {
      local($/);
      $html_string = <$data>;
    }
    else {
      $html_string = $data;
    }
    my %te_parms = (
      headers => ['Splits:'],
      debug   => $self->{debug},
    );
    my $te = HTML::TableExtract->new(%te_parms);
    $te->parse($html_string);
    my $table = $te->first_table_found || return [];
    my($split_line) = grep(defined && /split/i, $table->hrow);
    $split_line =~ s/^\s*splits:?\s*//i;
    my @rows;
    foreach (grep(/\w+/, split(/\]\s*,\s+/, $split_line))) {
      s/\s+$//;
      next if /none/i;
      next unless s/\s*\[(\d+):(\d+).*$//;
      my($post, $pre) = ($1, $2);
      my $date = ParseDate($_)
        or croak "Problem parsing date string '$_'\n";
      push(@rows, [$date, $post, $pre]);
    }
    \@rows;
  };
}

# Newer full-custom direct query for yahoo splits
sub _splits_old {
  my $self = shift;
  my @symbols = @_ ? @_ : $self->symbols;
  my $target_mode = 'split';
  # cache check
  my @not_seen;
  foreach my $symbol (@symbols) {
    my @r = $self->result_rows($target_mode, $symbol);
    push(@not_seen, $symbol) unless @r;
  }
  # example URL: http://finance.yahoo.com/q/bc?s=IBM&t=my
  foreach my $symbol (@not_seen) {
    my $url = $self->url_base_splits() . "?s=$symbol&t=my";
    print STDERR "Processing ($symbol:$target_mode) $url\n" if $self->{verbose};
    my $data = $self->{url_cache}{$url} || $self->fetch($url);
    $self->{url_cache}{$url} = $data;
    print STDERR "Custom parse for ($symbol:$target_mode)\n" if $self->{verbose};
    my $te = HTML::TableExtract->new(headers => ['Splits:'], debug => 5);
    print STDERR "DATA: |||\n$data\n|||\n";
    $te->parse($data);
    my $table = $te->first_table_found;
    print STDERR "TABLE: ", $table || 'undef', "\n";
    if ($table) {
      my($split_line) = grep(defined && /split/i, $table->hrow);
      $split_line =~ s/^\s*splits:?\s*//i;
      print STDERR "SPLIT LINE: ||| $split_line |||\n";
      my @rows;
      foreach (grep(/\w+/, split(/\]\s*,\s+/, $split_line))) {
        s/\s+$//;
        next if /none/i;
        next unless s/\s*\[(\d+):(\d+).*$//;
        my($post, $pre) = ($1, $2);
        my $date = ParseDate($_)
          or croak "Problem parsing date string '$_'\n";
        push(@rows, [$date, $post, $pre]);
      }
      @rows = $self->rows(\@rows);
      $self->_store_results($target_mode, $symbol, 0, \@rows);
      $self->_target_source($target_mode, $symbol, ref $self);
    }
  }
  my @rows = $self->result_rows($target_mode, @symbols);
  wantarray ? @rows : \@rows;
}

# Not so direct splits query
sub splits_alternate {
  # An HTML quote query is the only way to go for splits on yahoo, so
  # we have to signal to ourselves what the source mode should be so
  # that the _urls() method will generate the proper URL...otherwise
  # we would get CSV, which has no split info.
  my $self = shift;
  $self->save_query;
  $self->parse_mode('html');
  $self->start_date(undef, 0);
  $self->end_date('today', 0);
  $self->{quiet} = 1;
  $self->target_mode('dividend');
  my $rows = $self->dividends(@_);
  if (!@$rows) {
    $self->target_mode('quote');
    $self->granularity('monthly');
    $rows = $self->quotes(@_);
  }
  $self->restore_query;
  $self->result_rows('split');
}

sub labels {
  my $self = shift;
  my %parms = @_;
  my $target_mode = $self->target_mode;
  my @labels = $self->SUPER::labels(%parms);
  push(@labels, 'adj') if $target_mode eq 'quote';
  @labels;
}

sub extractors {
  # Override. If we are pulling CSV, save some time by skipping
  # extraction attempts
  my $self  = shift;
  my %parms = @_;
  my $target_mode = $parms{target_mode} || $self->target_mode;
  my $parse_mode  = $parms{parse_mode}  || $self->parse_mode;
  return () if $parse_mode eq 'csv';
  return () unless $target_mode eq 'quote' || $target_mode eq 'dividend';
  my $date_column = $self->label_column('date');
  my $split_column = 1;
  my %extractors;
  # for both quote and dividend results in html mode
  $extractors{'split'} = sub {
    my $row = shift;
    die "row as array ref required" unless ref $row;
    # example split: "3 : 1 Stock Split"
    my($post, $pre) = $row->[$split_column] =~ /(\d+)\s*:\s*(\d+).*Split/i;
    return undef unless $post && $pre;
    [ $row->[$date_column], $post, $pre ];
  };
  if ($target_mode eq 'quote') {
    my $div_column = 1;
    $extractors{dividend} = sub {
      # Get a row as array ref, see if it contains dividend info. If so,
      # return another array ref with the extracted info.
      my $row = shift;
      die "row as array ref required\n" unless ref $row;
      # example dividend: "$0.01 Cash Dividend"
      my($div) = $row->[$div_column] =~ /\$*(\d*\.\d+).*Dividend/i;
      return undef unless defined $div;
      [ $row->[$date_column], $row->[$div_column] ];
    };
  }
  %extractors;
}

sub url_maker {
  my($self, %parms) = @_;
  my $target_mode = $parms{target_mode} || $self->target_mode;
  my $parse_mode  = $parms{parse_mode}  || $self->parse_mode;

  # *always* block unknown target mode and parse mode combinations for
  # cascade to work properly!
  return undef unless $target_mode eq 'quote' ||
                      $target_mode eq 'split' ||
                      $target_mode eq 'dividend';
  return undef unless $parse_mode eq 'html' || $parse_mode eq 'csv';

  my $granularity = lc($parms{granularity} || $self->granularity);
  my $grain = 'd';
  $granularity =~ /^\s*(\w)/;
  $grain = $1 if $1 eq 'w' || $1 eq 'm';
  my($ticker, $start_date, $end_date) =
    @parms{qw(symbol start_date end_date)};
  $start_date ||= $self->start_date;
  $end_date   ||= $self->end_date;
  if ($start_date && $end_date && $start_date gt $end_date) {
    ($start_date, $end_date) = ($end_date, $start_date);
  }

  if ($target_mode eq 'split') {
    $self->parse_mode('html');
    my @urls = $self->url_base_splits() . "?s=$ticker&t=my";
    return sub { pop @urls };
  }

  my $base_url = $parse_mode eq 'csv' ? $self->url_base_csv() :
                                        $self->url_base_html();
  my @base_parms;
  if ($start_date) {
    my($y, $m, $d) = $self->ymd($start_date);
    $m = sprintf("%02d", $m - 1);
    push(@base_parms, "a=$m", "b=$d", "c=$y");
  }
  if ($end_date) {
    my($y, $m, $d) = $self->ymd($end_date);
    $m = sprintf("%02d", $m - 1);
    push(@base_parms, "d=$m", "e=$d", "f=$y");
  }
  my $g = $target_mode eq 'quote' ? $grain : 'v';
  $ticker ||= 'BOOLEAN';
  push(@base_parms, "g=$g", "s=$ticker");
  
  if ($parse_mode eq 'html' && $target_mode eq 'quote' || $target_mode eq 'dividend') {
    my $cursor = 0;
    my $window = 66;
    return sub {
      my $url = $base_url . '?' .
        join('&', @base_parms,
                 "z=$window", "y=$cursor");
      $url .= '&ignore=.csv' if $parse_mode eq 'csv';
      $cursor += $window;
      $url;
    }
  }
  else {
    my @urls = $base_url .  '?' . join('&', @base_parms);
    $urls[0] .= '&ignore=.csv' if $parse_mode eq 'csv';
    return sub { pop @urls }
  }
}

1;

__END__

=head1 NAME

Finance::QuoteHist::Yahoo - Site-specific subclass for retrieving historical stock quotes.

=head1 SYNOPSIS

  use Finance::QuoteHist::Yahoo;
  $q = new Finance::QuoteHist::Yahoo
     (
      symbols    => [qw(IBM UPS AMZN)],
      start_date => '01/01/2009',
      end_date   => 'today',
     );

  # Values
  foreach $row ($q->quotes()) {
    ($symbol, $date, $open, $high, $low, $close, $volume) = @$row;
    ...
  }

  # Splits
  foreach $row ($q->splits()) {
     ($symbol, $date, $post, $pre) = @$row;
  }

  # Dividends
  foreach $row ($q->dividends()) {
     ($symbol, $date, $dividend) = @$row;
  }

=head1 DESCRIPTION

Finance::QuoteHist::Yahoo is a subclass of
Finance::QuoteHist::Generic, specifically tailored to read historical
quotes, dividends, and splits from the Yahoo web site
(I<http://table.finance.yahoo.com/>).

For quotes and dividends, Yahoo can return data quickly in CSV format.
Both of these can also be extracted from HTML tables. Splits are
extracted from the HTML of the 'Basic Chart' page for that ticker.

There are no date range restrictions on CSV queries for quotes and
dividends.

For HTML queries, Yahoo takes arbitrary date ranges as arguments, but
breaks results into pages of 66 entries.

Please see L<Finance::QuoteHist::Generic(3)> for more details on usage
and available methods. If you just want to get historical quotes and
are not interested in the details of how it is done, check out
L<Finance::QuoteHist(3)>.

=head1 METHODS

The basic user interface consists of three methods, as seen in the
example above. Those methods are:

=over

=item quotes()

Returns a list of rows (or a reference to an array containing those
rows, if in scalar context). Each row contains the B<Symbol>, B<Date>,
B<Open>, B<High>, B<Low>, B<Close>, and B<Volume> for that date.

=item dividends()

Returns a list of rows (or a reference to an array containing those
rows, if in scalar context). Each row contains the B<Symbol>, B<Date>,
and amount of the B<Dividend>, in that order.

=item splits()

Returns a list of rows (or a reference to an array containing those
rows, if in scalar context). Each row contains the B<Symbol>, B<Date>,
B<Post> split shares, and B<Pre> split shares, in that order.

=back

The following methods override methods provided by the
Finance::QuoteHist::Generic module; more of this was necessary than is
typical for a basic query site due to the variety of query types and
data formats available on Yahoo.

=over

=item url_maker()

Returns a subroutine reference tailored for the current target mode and
parsing mode. The routine is an iterator that will produce all necessary
URLs on repeated invocations necessary to complete a query.

=item extractors()

Returns a hash of subroutine references that attempt to extract embedded
values (dividends or splits) within the results from a larger query.

=item labels()

Includes the 'adj' column.

=back

=head1 REQUIRES

Finance::QuoteHist::Generic

=head1 DISCLAIMER

The data returned from these modules is in no way guaranteed, nor are
the developers responsible in any way for how this data (or lack
thereof) is used. The interface is based on URLs and page layouts that
might change at any time. Even though these modules are designed to be
adaptive under these circumstances, they will at some point probably
be unable to retrieve data unless fixed or provided with new
parameters. Furthermore, the data from these web sites is usually not
even guaranteed by the web sites themselves, and oftentimes is
acquired elsewhere.

If you would like to know more, check out the terms of service from
Yahoo!, which can be found here:

  http://docs.yahoo.com/info/terms/

If you still have concerns, then use another site-specific historical
quote instance, or none at all.

Above all, play nice.

=head1 AUTHOR

Matthew P. Sisk, E<lt>F<sisk@mojotoad.com>E<gt>

=head1 COPYRIGHT

Copyright (c) 2000-2010 Matthew P. Sisk. All rights reserved. All wrongs
revenged. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

Finance::QuoteHist::Generic(3), Finance::QuoteHist(3), perl(1).

=cut
